qcodes_contrib_drivers/drivers/SingleQuantum/SingleQuantum.py

```diff
@@ -0,0 +1,498 @@
+        self.TCP_IP_ADR = TCP_IP_ADR
+        self.TCP_IP_PORT = TCP_IP_PORT
+
+        self.socket = socket.socket(
+            socket.AF_INET, socket.SOCK_STREAM)
+        self.socket.connect((self.TCP_IP_ADR, self.TCP_IP_PORT))
+        # self.socket.settimeout(.1)
+        self.BUFFER = 1000000
+        self.shutdown = False
+
+        self.cnts = []
+        self.CNTS_BUFFER = CNTS_BUFFER
+        self.n = 0
+
+    @synchronized_method
+    def close(self):
+        # print("Closing Socket")
+        self.socket.close()
+        self.shutdown = True
+
+    @synchronized_method
+    def get_n(self, n):
+        n0 = self.n
+        while self.n < n0 + n:
+            time.sleep(0.001)
+        cnts = self.cnts
+        return cnts[-n:]
+
+    def run(self):
+        while self.shutdown is False:
+            if sys.version_info.major == 3:
+                data_raw = str(self.socket.recv(self.BUFFER), 'utf-8')
+            elif sys.version_info.major == 2:
+                data_raw = self.socket.recv(self.BUFFER)
+
+            data_newline = data_raw.split('\n')
+
+            v = []
+            for d in data_newline[0].split(','):
+                v.append(float(d))
+
+            with self.lock:
+                self.cnts.append(v)
+                # Keep Size of self.cnts
+                len_ = len(self.cnts)
+                if len_ > self.CNTS_BUFFER:
+                    self.cnts = self.cnts[len_ - self.CNTS_BUFFER:]
+                self.n += 1
+
+
+class ChannelArray(ParameterWithSetpoints):
+    """Fetches the correct row from the matrix counters,  based on the channel
+    This is a parameter with setpoints, where the setpoints are the time stamps
+    Args:
+        channel (int): the channel number to select the row from the counters matrix
+    Return (numpy_array): a numpy vector of length npts, containing the counts
+    """
+
+    def __init__(self, channel, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._channel = channel
+
+    def get_raw(self):
+        return self.root_instrument.counters.get_latest()[
+            self._channel]
+
+
+class TimeArray(Parameter):
+    """Fetches the row with time stamps, currently channel 0
+    This parameter is used as the setpoints for the channelarray
+    Args:
+        channel (int): the channel number to select the row from the counters matrix (channel 0 is the timestamp)
+    Return (numpy_array): a numpy vector of length npts, containing the time stamps
+    """
+
+    def __init__(self, channel, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._channel = channel
+
+    def get_raw(self):
+        return np.array(self.root_instrument.counters.get_latest()[
+            self._channel])
+
+
+class WebSQControlqcode(IPInstrument):
+    """The instrument.
+    The first part are functions to talk in JSON style with the instrument.
+    The second part contains QCoDeS init and parameters.
```

Comment:
Since this is split in 2 parts I would considder putting all of the first part in a mixin class for it self

Reply:
Done.